File: LeetCode_Algorithm/56(array_sort)/56.py
from typing import List

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        # sol1 O(nlogn) 218ms 36%
        intervals.sort(key=lambda x : x[0])
        ans = []
        for interval in intervals:
            if ans == [] or ans[-1][1] < interval[0]:
                ans.append(interval)
            else:
                ans[-1][1] = max(ans[-1][1], interval[1])
        return ans
            

# print(Solution().merge([[1,3],[2,6],[8,10],[15,18]]))
# print(Solution().merge([[1,4],[0,4]]))
# print(Solution().merge([[1,4],[0,0]]))
# print(Solution().merge([[2,3],[4,5],[6,7],[8,9],[1,10]]))
# print(Solution().merge([[1,4], [1,4]]))
print(Solution().merge([[1,4], [2,3]]))

File: LeetCode_Algorithm/56(array_sort)/56.js
var merge = function(intervals) {
  // sol1 O(nlogn) 177ms 37%
  if (intervals.length == 0) {return []}
  intervals.sort((a,b)=>{
    if(a[0] == b[0]) {
      return a[1]-b[1]
    } else {
      return a[0]-b[0]
    }
  })
  
  let i = 0; let ans = []
  while (i < intervals.length) {
    if (ans.length == 0 || intervals[i][0] > ans[ans.length-1][1]) {
      ans.push(intervals[i])
    } else {
      ans[ans.length-1][1] = Math.max(ans[ans.length-1][1], intervals[i][1])
    }
    i += 1
  }
  return ans
};

console.log(merge([[1,13],[2,6],[1,10],[15,18]]))